File: Algorithms/0052_N-Queens_II/Python/N-Queens_II_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def totalNQueens(self, n):
        self.queen_position = set()
        self.res = 0

        def backtracking(row_num, queen_count, required_queen, length):
            if queen_count == required_queen:
                self.res += 1
                return

            for x in range(length):
                if self.check_position(x, row_num, length, self.queen_position):
                    self.queen_position.add((row_num, x))
                    backtracking(row_num + 1, queen_count + 1, required_queen, length)
                    self.queen_position.remove((row_num, x))

        backtracking(0, 0, n, n)

        return self.res

    def check_position(self, x, y, length, alist):
        if len(alist) == 0: return True

        for queen_y, queen_x in alist:
            if x == queen_x or y == queen_y: return False

        direction = [[1, 1], [-1, -1], [-1, 1], [1, -1]]

        for dir_y, dir_x in direction:
            temp_x, temp_y = x, y

            while 0 <= temp_x < length and 0 <= temp_y < length:
                if (temp_y, temp_x) in alist: return False
                temp_x += dir_x
                temp_y += dir_y

        return True

File: Algorithms/0052_N-Queens_II/Java/N-QueensIISolution1.java


// Space: O(n)
// Time: O(n)

class Solution {
    int res = 0;

    public int totalNQueens(int n) {
        if (n <= 1) return n;
        String[][] board = new String[n][n];
        for (String[] row : board) {
            Arrays.fill(row, ".");
        }
        dfs(board, n, 0);
        return res;
    }

    private void dfs(String[][] board, int queens, int row) {
        if (queens == 0) {
            res += 1;
            return;
        }
        for (int i = 0; i < board.length; i++) {
            if (verify(board, row, i)) {
                board[row][i] = "Q";
                dfs(board, queens - 1, row + 1);
                board[row][i] = ".";
            }
        }
    }

    private boolean verify(String[][] board, int y, int x) {
        int temp_y = y;
        int temp_x = x;
        while (temp_y >= 0) {
            if (board[temp_y][x] != ".") return false;
            temp_y -= 1;
        }
        temp_y = y;
        while (temp_y >= 0 && temp_x >= 0) {
            if (board[temp_y][temp_x] != ".") return false;
            temp_y -= 1;
            temp_x -= 1;
        }
        temp_y = y;
        temp_x = x;
        while (temp_y >= 0 && temp_x < board.length) {
            if (board[temp_y][temp_x] != ".") return false;
            temp_y -= 1;
            temp_x += 1;
        }
        return true;
    }
}




